bin/fastq_scan.py

```diff
@@ -1,237 +1,299 @@
-        "read_qual_std",
-        "mean_sequence_length",
-        "min_sequence_length",
-        "max_sequence_length",
-        "std_sequence_length",
-    ]
-    total_bp: int
-    total_reads: int
-    qual_min: Union[float, decimal.Decimal]
-    qual_max: Union[float, decimal.Decimal]
-    qual_sum: int
-    qual_mean: Union[float, decimal.Decimal]
-    qual_std: Union[float, decimal.Decimal]
-    read_qual_mean: Union[float, decimal.Decimal]
-    read_qual_std: Union[float, decimal.Decimal]
-    mean_sequence_length: Union[float, decimal.Decimal]
-    min_sequence_length: Union[float, decimal.Decimal]
-    max_sequence_length: Union[float, decimal.Decimal]
-    std_sequence_length: Union[float, decimal.Decimal]
-
-    # def __repr__(self):
-    #    # TODO just call asdict on output
-    #    repr_val = "\n\t{" + f"""
-    #    \t\ttotal_bp: {self.total_bp},
-    #    \t\ttotal_reads: {self.total_reads},
-    #    \t\tqual_min: {int(self.qual_min)},
-    #    \t\tqual_max: {int(self.qual_max)},
-    #    \t\tqual_sum: {int(self.qual_sum)},
-    #    \t\tqual_mean: {round(self.qual_mean, 2)},
-    #    \t\tqual_std: {round(self.qual_std, 2)},
-    #    \t\tread_qual_mean: {round(self.read_qual_mean, 2)},
-    #    \t\tread_qual_std: {round(self.read_qual_std, 2)},
-    #    \t\tmean_sequence_length: {round(self.mean_sequence_length, 2)},
-    #    \t\tstd_sequence_length: {round(self.mean_sequence_length, 2)},
-    #    \t\tmin_sequence_length: {int(self.min_sequence_length)},
-    #    \t\tmax_sequence_length: {int(self.max_sequence_length)},
-    #    """ +"\t}\n"
-    #    return repr_val
-
-
-class FastQReader:
-    """A slow but accurate program for calculating simple fastq metrics
-
-    Returns:
-        _type_: _description_
+
+GZIP_MAGIC_NUMBERS = b"\x1f\x8b"
+FQ_EXTENSIONSj = frozenset([".FASTQ", ".FQ"])
+
+GZIP_READER = partial(gzip.open, mode="rt")
+STD_READER = partial(open, mode="r")
+
+EXPECTED_LINES_IN_FASTQ = 4
+
+
+def get_file_reader(file: p.Path) -> io.StringIO:
+    """
+    Return a text stream based on a files encoding. File is assumed to be plaintext
+    unless there is information showing that it is compressed. As the FastQ format
+    will be validated downstream as it is not easy determining if a file is
+    plaintext or binary.
+    """
+
+    f_reader = STD_READER
+    with open(file, "rb") as test_f:
+        test_bytes = test_f.read(2)
+        if test_bytes == GZIP_MAGIC_NUMBERS:
+            f_reader = GZIP_READER
+    return f_reader(file)
+
+
+class FastQData:
+    """
+    This object is used for tracking FastQ data from a file as
+    it is read. Counters are stored of quality data in the form of a
+    bit vector as the number of times an observation is seen can be tracked.
+
+    Sequence lenghts will be tracked as a dictionary with counts denoting a given
+    read size, this will prevent the need of using an exceedingly large list from
+    being kept. In reality this could be kept in a BST
```

Comment:
I don't know if it would make sense for this to be a binary search tree. A dictionary will have a faster lookup time. A BST might make sense if space is a problem, but a dictionary is probably easiest here.